crates/lsp/src/handlers_state.rs

```diff
@@ -199,6 +191,36 @@ pub(crate) async fn did_change_configuration(
         .await
 }
 
+pub(crate) fn did_change_workspace_folders(
+    params: DidChangeWorkspaceFoldersParams,
+    lsp_state: &mut LspState,
+) -> anyhow::Result<()> {
+    // Collect all `errors` to ensure we don't drop events after a first error
+    let mut errors = ErrorVec::new();
+
+    for lsp_types::WorkspaceFolder { uri, .. } in params.event.added {
+        errors.push_err(lsp_state.open_workspace_folder(&uri, Settings::default()));
+    }
+    for lsp_types::WorkspaceFolder { uri, .. } in params.event.removed {
+        errors.push_err(lsp_state.close_workspace_folder(&uri));
+    }
+
+    errors.into_result()
+}
```

Comment:
Updating our per-workspace settings resolver when a workspace is opened / closed